### PR TITLE
ECSOperator: airflow exception on edge case when cloudwatch log stream is not found

### LINKED ISSUE #15000: When an ECS Task fails to start, ECS Operator raises a CloudWatch exception
**Apache Airflow version**: 1.10.13

**Environment**:

- **Cloud provider or hardware configuration**:AWS
- **OS** (e.g. from /etc/os-release): Amazon Linux 2
- **Kernel** (e.g. `uname -a`):  4.14.209-160.339.amzn2.x86_64
- **Install tools**: pip
- **Others**:

**What happened**:

When an ECS Task exits with `stopCode: TaskFailedToStart`, the ECS Operator will exit with a ResourceNotFoundException for the GetLogEvents operation. This is because the task has failed to start, so no log is created.

```
[2021-03-14 02:32:49,792] {ecs_operator.py:147} INFO - ECS Task started: {'tasks': [{'attachments': [], 'availabilityZone': 'ap-northeast-1c', 'clusterArn': 'arn:aws:ecs:ap-northeast-1:xxxx:cluster/ecs-cluster', 'containerInstanceArn': 'arn:aws:ecs:ap-northeast-1:xxxx:container-instance/ecs-cluster/xxxx', 'containers': [{'containerArn': 'arn:aws:ecs:ap-northeast-1:xxxx:container/xxxx', 'taskArn': 'arn:aws:ecs:ap-northeast-1:xxxx:task/ecs-cluster/xxxx', 'name': 'container_image', 'image': 'xxxx.dkr.ecr.ap-northeast-1.amazonaws.com/ecr/container_image:latest', 'lastStatus': 'PENDING', 'networkInterfaces': [], 'cpu': '128', 'memoryReservation': '128'}], 'cpu': '128', 'createdAt': datetime.datetime(2021, 3, 14, 2, 32, 49, 770000, tzinfo=tzlocal()), 'desiredStatus': 'RUNNING', 'group': 'family:task', 'lastStatus': 'PENDING', 'launchType': 'EC2', 'memory': '128', 'overrides': {'containerOverrides': [{'name': 'container_image', 'command': ['/bin/bash', '-c', 'xxxx']}], 'inferenceAcceleratorOverrides': []}, 'startedBy': 'airflow', 'tags': [], 'taskArn': 'arn:aws:ecs:ap-northeast-1:xxxx:task/ecs-cluster/xxxx', 'taskDefinitionArn': 'arn:aws:ecs:ap-northeast-1:xxxx:task-definition/task:1', 'version': 1}], 'failures': [], 'ResponseMetadata': {'RequestId': 'xxxx', 'HTTPStatusCode': 200, 'HTTPHeaders': {'x-amzn-requestid': 'xxxx', 'content-type': 'application/x-amz-json-1.1', 'content-length': '1471', 'date': 'Sun, 14 Mar 2021 02:32:48 GMT'}, 'RetryAttempts': 0}}
[2021-03-14 02:34:15,022] {ecs_operator.py:168} INFO - ECS Task stopped, check status: {'tasks': [{'attachments': [], 'availabilityZone': 'ap-northeast-1c', 'clusterArn': 'arn:aws:ecs:ap-northeast-1:xxxx:cluster/ecs-cluster', 'connectivity': 'CONNECTED', 'connectivityAt': datetime.datetime(2021, 3, 14, 2, 32, 49, 770000, tzinfo=tzlocal()), 'containerInstanceArn': 'arn:aws:ecs:ap-northeast-1:xxxx:container-instance/ecs-cluster/xxxx', 'containers': [{'containerArn': 'arn:aws:ecs:ap-northeast-1:xxxx:container/xxxx', 'taskArn': 'arn:aws:ecs:ap-northeast-1:xxxx:task/ecs-cluster/xxxx', 'name': 'container_image', 'image': 'xxxx.dkr.ecr.ap-northeast-1.amazonaws.com/ecr/container_image:latest', 'lastStatus': 'STOPPED', 'reason': 'CannotPullContainerError: failed to register layer: Error processing tar file(exit status 1): write /var/lib/xxxx: no space left on device', 'networkInterfaces': [], 'healthStatus': 'UNKNOWN', 'cpu': '128', 'memoryReservation': '128'}], 'cpu': '128', 'createdAt': datetime.datetime(2021, 3, 14, 2, 32, 49, 770000, tzinfo=tzlocal()), 'desiredStatus': 'STOPPED', 'executionStoppedAt': datetime.datetime(2021, 3, 14, 2, 34, 12, 810000, tzinfo=tzlocal()), 'group': 'family:task', 'healthStatus': 'UNKNOWN', 'lastStatus': 'STOPPED', 'launchType': 'EC2', 'memory': '128', 'overrides': {'containerOverrides': [{'name': 'container_image', 'command': ['/bin/bash', '-c', 'xxxx']}], 'inferenceAcceleratorOverrides': []}, 'pullStartedAt': datetime.datetime(2021, 3, 14, 2, 32, 51, 68000, tzinfo=tzlocal()), 'pullStoppedAt': datetime.datetime(2021, 3, 14, 2, 34, 13, 584000, tzinfo=tzlocal()), 'startedBy': 'airflow', 'stopCode': 'TaskFailedToStart', 'stoppedAt': datetime.datetime(2021, 3, 14, 2, 34, 12, 821000, tzinfo=tzlocal()), 'stoppedReason': 'Task failed to start', 'stoppingAt': datetime.datetime(2021, 3, 14, 2, 34, 12, 821000, tzinfo=tzlocal()), 'tags': [], 'taskArn': 'arn:aws:ecs:ap-northeast-1:xxxx:task/ecs-cluster/xxxx', 'taskDefinitionArn': 'arn:aws:ecs:ap-northeast-1:xxxx:task-definition/task:1', 'version': 2}], 'failures': [], 'ResponseMetadata': {'RequestId': 'xxxx', 'HTTPStatusCode': 200, 'HTTPHeaders': {'x-amzn-requestid': 'xxxx', 'content-type': 'application/x-amz-json-1.1', 'content-length': '1988', 'date': 'Sun, 14 Mar 2021 02:34:14 GMT'}, 'RetryAttempts': 0}}
[2021-03-14 02:34:15,024] {ecs_operator.py:172} INFO - ECS Task logs output:
[2021-03-14 02:34:15,111] {credentials.py:1094} INFO - Found credentials in environment variables.
[2021-03-14 02:34:15,416] {taskinstance.py:1150} ERROR - An error occurred (ResourceNotFoundException) when calling the GetLogEvents operation: The specified log stream does not exist.
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/airflow/models/taskinstance.py", line 984, in _run_raw_task
    result = task_copy.execute(context=context)
  File "/usr/local/lib/python3.7/site-packages/airflow/contrib/operators/ecs_operator.py", line 152, in execute
    self._check_success_task()
  File "/usr/local/lib/python3.7/site-packages/airflow/contrib/operators/ecs_operator.py", line 175, in _check_success_task
    for event in self.get_logs_hook().get_log_events(self.awslogs_group, stream_name):
  File "/usr/local/lib/python3.7/site-packages/airflow/contrib/hooks/aws_logs_hook.py", line 85, in get_log_events
    **token_arg)
  File "/usr/local/lib/python3.7/site-packages/botocore/client.py", line 357, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/usr/local/lib/python3.7/site-packages/botocore/client.py", line 676, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.errorfactory.ResourceNotFoundException: An error occurred (ResourceNotFoundException) when calling the GetLogEvents operation: The specified log stream does not exist.
```

**What you expected to happen**:

ResourceNotFoundException is misleading because it feels like a problem with CloudWatchLogs. Expect AirflowException to indicate that the task has failed.

**How to reproduce it**:

This can be reproduced by running an ECS Task that fails to start, for example by specifying a non-existent entry_point.

**Anything else we need to know**:

I suspect Issue #11663 has the same problem, i.e. it's not a CloudWatch issue, but a failure to start an ECS Task.

### PR DESCRIPTION
**Pull request by pavelhlushchanka**
Closes #15000, related to PR #17626

In PR #17626 there is a small bug that doesn't handle the case of #15000 at all. So the fix handles properly #15000 and fixes the bug introduced in #17626.

### CODE REVIEW

**Comment by codenamestif** [comment]
@zachliu @potiuk could you please review it? Thanks.

**Comment by codenamestif** [comment]
@zachliu i'm double-checking the issue #15000. So would it be better to check `stopCode: TaskFailedToStart` rather than existence of the log stream of the task in order to handle the case properly? Because in #17626 you said that log stream is a kinda flag when ECS fails to provision tasks.

**Review by potiuk — APPROVED** [review]

**Comment by zachliu** [comment]
> @zachliu i'm double-checking the issue #15000. So would it be better to check `stopCode: TaskFailedToStart` rather than existence of the log stream of the task in order to handle the case properly? Because in #17626 you said that log stream is a kinda flag when ECS fails to provision tasks.

well ideally it would be better. but since that thing is hiding deep in the json response, the code may not look so good :joy_cat: 

```
response.get('tasks')[0].get('stoppedReason')
response.get('tasks')[0].get('stopCode')
```

**Comment by codenamestif** [comment]
> > @zachliu i'm double-checking the issue #15000. So would it be better to check `stopCode: TaskFailedToStart` rather than existence of the log stream of the task in order to handle the case properly? Because in #17626 you said that log stream is a kinda flag when ECS fails to provision tasks.
> 
> well ideally it would be better. but since that thing is hiding deep in the json response, the code may not look so good 😹
> 
> ```
> response.get('tasks')[0].get('stoppedReason')
> response.get('tasks')[0].get('stopCode')
> ```

What if cloudwatch logs are disabled in the operator (group/streams are not provided), then the edge case would not be caught since we don't know what log stream to check, right? Then the only way is to really parse the json.

File: airflow/providers/amazon/aws/operators/ecs.py [hunk 1]
```diff
@@ -452,6 +460,11 @@ def _check_success_task(self) -> None:
         if len(response.get('failures', [])) > 0:
             raise AirflowException(response)
 
+        if self.task_log_fetcher and not self.task_log_fetcher.is_log_stream_created():
+            raise AirflowException(
+                'Cloudwatch log stream for the task is not found. The task probably failed to start.'
+            )
+
         for task in response['tasks']:
```
Comment (pavelhlushchanka) [inline airflow/providers/amazon/aws/operators/ecs.py:455]:
```suggestion
        for task in response['tasks']:
            if task.get('stopCode', '') == 'TaskFailedToStart':
                raise AirflowException('The task failed to start.')
```

Such check is probably better and simpler. And no need to check cloudwatch log streams.

**Comment by zachliu** [comment]
> > > @zachliu i'm double-checking the issue #15000. So would it be better to check `stopCode: TaskFailedToStart` rather than existence of the log stream of the task in order to handle the case properly? Because in #17626 you said that log stream is a kinda flag when ECS fails to provision tasks.
> > 
> > 
> > well ideally it would be better. but since that thing is hiding deep in the json response, the code may not look so good joy_cat
> > ```
> > response.get('tasks')[0].get('stoppedReason')
> > response.get('tasks')[0].get('stopCode')
> > ```
> 
> What if cloudwatch logs are disabled in the operator (group/streams are not provided), then the edge case would not be caught since we don't know what log stream to check, right? Then the only way is to really parse the json.

i was just joking around AWS's response structure :grin:
yeah parsing the json is the ultimate solution :+1:

File: airflow/providers/amazon/aws/operators/ecs.py (hunk 1 above)
Comment (zachliu) [reply airflow/providers/amazon/aws/operators/ecs.py:455]: how about
```python
raise AirflowException('The task failed to start due to: {}'.format(task.get('stoppedReason', '')))
```

File: airflow/providers/amazon/aws/operators/ecs.py [hunk 2]
```diff
@@ -452,6 +460,11 @@ def _check_success_task(self) -> None:
...
             # This is a `stoppedReason` that indicates a task has not
             # successfully finished, but there is no other indication of failure
```
Comment (zachliu) [inline airflow/providers/amazon/aws/operators/ecs.py:461]: aha, we do check `stoppedReason` here. but the "Host EC2 (instance id) stopped/terminated" is a much rarer reason than "Timeout waiting for network interface provisioning to complete." according to AWS support when i asked about "common" reasons for silently failed tasks

Comment (pavelhlushchanka) [reply airflow/providers/amazon/aws/operators/ecs.py:461]: Yes, but in our case we have a clear `stopCode` that has to be handled. So it's easier. In that place we can have some edge case handlers, i think that's totally fine.

Comment (zachliu) [reply airflow/providers/amazon/aws/operators/ecs.py:461]: :+1:

**Comment by codenamestif** [comment]
It should be good to go now.

**Comment by potiuk** [comment]
Or check-out main version nd run it there (with Breeze providers are by default installed from local sources)

**Comment by zachliu** [comment]
> Or check-out main version nd run it there (with Breeze providers are by default installed from local sources)

oh, so i cannot just test this provider with my 2.1.4?

**Comment by uranusjr** [comment]
You can

```
# Check out the latest main branch...

# This is basically how the actual release is built, sans the upload part.
./breeze prepare-provider-packages --package-format=wheel --version-suffix-for-pypi=rc2 amazon

# Replace this with the correct pip to your Airflow 2.1.4 installation.
pip install ./dist/apache_airflow_providers_amazon-2.3.0rc2-py3-none-any.whl
```

**Comment by zachliu** [comment]
`prepare-provider-packages`! how do i forget this...

oh-right-yeah

**Comment by codenamestif** [comment]
@zachliu could you please confirm in #18835 that the fix is working fine for you? Thanks.

**Comment by zachliu** [comment]
> @zachliu could you please confirm in #18835 that the fix is working fine for you? Thanks.

the fix has been in production for 4~5 days now. funny thing is AWS must have upgraded something in the backend, the `TaskFailedToStart` occurs so rarely these days :joy_cat: 

in Aug, it usually happened to 2~4% of our daily tasks
now in Oct, the rate dropped to 0.07% (got this number before i applied #18835)

haven't seen a single `AirflowException('The task failed to start due to: {}'.format(task.get('stoppedReason', '')))` yet

**Pull request merged**